### PR TITLE
Add API to generate kubeconfig for a serviceAccount

### PR DESCRIPTION
**Pull request by yaocw2020**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:** Harvester cloud provider needs a kubeconfig

**Solution:** Add an API to generate kubeconfig for a serviceAccount

**Related Issue:** https://github.com/harvester/harvester/issues/1019

**Test plan:**

### CODE REVIEW

**Comment by yaocw2020** [comment]
API path:  `/v1/harvester/kubeconfig`
Request body:
``` json
{
    "serverUrl": "https://172.16.0.203:6443",
    "clusterRoleName": "harvesterhci.io:cloudprovider",
    "namespace": "default",
    "serviceAccountName": "abcd"
}
```
- `serverUrl`: The URL to access apiserver.
- `clusterRoleName`: The name of the created clusterRole.
- `namespace`: The namespace where to create the serviceAccount.
- `serviceAccountName`: The name of the serviceAccount to be created. 

The backend will create a serviceAccount in the specified namespace and a new clusterRoleBinding to bind the serviceAccount and clusterRole.

**Comment by gitlawr** [comment]
1. We need to ensure that this API can only be called by users with sa/rolebinding creation permissions.
2. Is there a way to clean up the underlying resources(sa/rolebinding) automatically?

**Comment by yaocw2020** [comment]
> 1. We need to ensure that this API can only be called by users with sa/rolebinding creation permissions.
> 2. Is there a way to clean up the underlying resources(sa/rolebinding) automatically?



> 1. We need to ensure that this API can only be called by users with sa/rolebinding creation permissions.
> 2. Is there a way to clean up the underlying resources(sa/rolebinding) automatically?

1. T

**Comment by yaocw2020** [comment]
> 1. We need to ensure that this API can only be called by users with sa/rolebinding creation permissions.
> 2. Is there a way to clean up the underlying resources(sa/rolebinding) automatically?

1. TO do it after integration with Rancher.
2. No, we can not delete them automatically.

File: pkg/api/kubeconfig/generate_handler.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+		secretClient:             scaled.CoreFactory.Core().V1().Secret(),
+	}
+}
+
+type req struct {
+	ServerUrl       string `json:"serverUrl"`
+	ClusterRoleName string `json:"clusterRoleName"`
+	Namespace       string `json:"namespace"`
+	SaName          string `json:"serviceAccountName"`
+}
```
Comment (guangbochen) [inline pkg/api/kubeconfig/generate_handler.go:59]: can we add unit testing for validating the request input?

File: pkg/api/kubeconfig/generate_handler.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+func (h *GenerateHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	// TODO authentication
+	var req req
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		util.ResponseError(rw, http.StatusBadRequest, errors.Wrap(err, "fail to decode request"))
+		return
+	}
```
Comment (guangbochen) [inline pkg/api/kubeconfig/generate_handler.go]: should we add not empty input validation?

File: pkg/api/kubeconfig/generate_handler.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+		util.ResponseError(rw, http.StatusInternalServerError, errors.Wrap(err, "fail to generate kubeconfig"))
+		return
+	}
+
+	util.ResponseOKWithBody(rw, kubeConfig)
+}
+
+func (h *GenerateHandler) setClusterRole(clusterRoleName string) error {
+	var err error
+	h.clusterRole, err = h.clusterRoleCache.Get(clusterRoleName)
```
Comment (gitlawr) [inline pkg/api/kubeconfig/generate_handler.go]: Don't set clusterRole as a field of the handler.

File: pkg/api/kubeconfig/generate_handler.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+	return err
+}
+
+func (h *GenerateHandler) createClusterRoleBindingIfNotExit(clusterRoleName string, sa *corev1.ServiceAccount) (*rbacv1.ClusterRoleBinding, error) {
```
Comment (gitlawr) [inline pkg/api/kubeconfig/generate_handler.go]: We should create rolebinding, not clusterrolebinding.

File: pkg/api/kubeconfig/generate_handler.go [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "ClusterRole",
+			Name:     clusterRoleName,
+		},
+	})
+}
+
+// createSaIfNotExit returns the serviceAccount and the associated secret
+func (h *GenerateHandler) ensureSaAndSecret(namespace, name string) (*corev1.ServiceAccount, *corev1.Secret, error) {
```
Comment (gitlawr) [inline pkg/api/kubeconfig/generate_handler.go]: The comment does not match the function name, and has a typo

File: pkg/api/kubeconfig/generate_handler.go [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+		return "", nil
+	}
+
+	return b.String(), nil
+}
+
+func (h *GenerateHandler) getCaAndToken(secret *corev1.Secret) ([]byte, []byte, error) {
+	ca, ok := secret.Data["ca.crt"]
+	if !ok {
+		return nil, nil, fmt.Errorf("not find ca.crt in secret %s", secret.Name)
```
Comment (gitlawr) [inline pkg/api/kubeconfig/generate_handler.go]: ca.crt is not found

File: pkg/api/kubeconfig/generate_handler.go [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+	}
+	token, ok := secret.Data["token"]
+	if !ok {
+		return nil, nil, fmt.Errorf("not find token in secret %s", secret.Name)
```
Comment (gitlawr) [inline pkg/api/kubeconfig/generate_handler.go]: ditto

File: pkg/api/kubeconfig/generate_handler.go [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+		Cluster:   cluster,
+		Namespace: secret.Namespace,
+		AuthInfo:  user,
+	}
+	config.CurrentContext = context
+
+	b := new(strings.Builder)
+	if err := clientcmdlatest.Codec.Encode(config, b); err != nil {
+		return "", nil
+	}
```
Comment (gitlawr) [inline pkg/api/kubeconfig/generate_handler.go]: Why don't we return the error

File: pkg/server/router.go [hunk 9]
```diff
@@ -78,6 +79,9 @@ func (r *Router) Routes(h router.Handlers) http.Handler {
 	sbDownloadHandler := supportbundle.NewDownloadHandler(r.scaled, r.options.Namespace)
 	m.Path("/v1/supportbundles/{bundleName}/download").Methods("GET").Handler(sbDownloadHandler)
 
+	kcGenerateHandler := kubeconfig.NewGenerateHandler(r.scaled)
+	m.Path("/v1/kubeconfig").Methods("POST").Handler(kcGenerateHandler)
```
Comment (gitlawr) [inline pkg/server/router.go]: We need to access via the `/v1/harvester` aggregation APIs

Comment (yaocw2020) [reply pkg/server/router.go]: I tried the path `/v1/harvester/kubeconfig`, but got 404.

**Review by guangbochen — APPROVED** [review]
LGTM

**Review by gitlawr — APPROVED** [review]
LGTM

**Pull request merged**